### ISSUE TITLE
Refactor to remove `fixed` from `FixerData`

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#8233]
see: https://github.com/stylelint/stylelint/blob/3751cfb4c6ba8c94dbd88db75ff6bd77a1aa70bc/lib/utils/report.mjs#L168-L202

While working on https://github.com/stylelint/stylelint/issues/7192 I found that the `fixed` flag in `FixerData` is confusing and limiting.

I assume the intention was to be able to differentiate between a fix that was purely run to compute the replacement text and a fix that was also applied on the final output.

However I don't think there is a use case for that.

Either users want all fixes to be applied by Stylelint or they want no fixes to be applied at all.
Optionally users might want a report of potential fixes that could have been applied.

But I don't think users ever want both interleaved, or am I overlooking something?

------

I suggest we remove this flag and refactor `report` to only record fixes in `fixersData`.

i.e.

```js
/** @param {Problem & { line: number }} problem */
function isFixApplied({ fix, line, result: { stylelint }, ruleName }) {
	const { disabledRanges, config = {}, fixersData } = stylelint;

	if (!isFn(fix)) {
		return false;
	}

	const shouldFix = Boolean(config.fix && !config.rules?.[ruleName][1]?.disableFix);
	const mayFix =
		shouldFix && (config.ignoreDisables || !isDisabled(ruleName, line, disabledRanges));

	if (!mayFix) return false;

	fix();
	addFixData({ fixersData, ruleName });

	return true;
}
```

_This also requires refactoring the `verbose` formatter so that it simply reports all fixers data. Currently it filters entries where `fixed == true`._

-----

To resolve https://github.com/stylelint/stylelint/issues/7192 we will eventually have to calculate replacement text, which will likely require running `fix()`.

By the current implementation this would set the `fixed` flag to true.

But for https://github.com/stylelint/stylelint/issues/7192 it is also important that no fixers are actually applied.

I think the fixed flag is mainly ambiguous in its meaning:
- a fix was computed?
- a fix was applied to your source files?

Removing it altogether is a simple way to avoid the ambiguity.

-------

I do think there might a need for several other flags:
- were some fixes skipped because of configuration comments?
- are there any fixes left that couldn't be computed in a first pass?

**Comment by romainmenke** [comment]
see : https://github.com/stylelint/stylelint/blob/cf17793d6e4bd159352d78d142cbcda645ee952a/lib/utils/report.mjs#L178-L221

I am not sure what the exact expectations are around the `FixerData`, so I might have mutated it too much.

As I currently understand it there is a use case for knowing what was fixed when autofixes are applied. So that end users might have some feedback on what was changed. (see https://github.com/stylelint/stylelint/issues/4664)

With that in mind I altered the recording of the `FixerData`.

It now only adds an entry when `fix()` was called in a context where it would (likely) affect the original source file.

It does not add an entry when:
- `fix()` was not called (fix disabled for a rule, configuration comments, ...)
- `config.fix` **is not** truthy
- `config.computeReplacementText` **is** truthy

That PR is only a draft and we likely want to split up those changes into smaller chunks.
The changes in https://github.com/stylelint/stylelint/pull/8234 simply made it even more unclear when and what data should be added as `FixerData`

**Comment by Mouvedia** [comment]
> To resolve #7192 we will eventually have to calculate replacement text, which will likely require running `fix()`.

In the discussion starting around https://github.com/stylelint/stylelint/issues/7784#issuecomment-2184094675 many proposals were mentioned:

```ts
interface Fix {
	(): void | never;
	editInfo?: EditInfo;
}

export type Problem = {
	/* … */
	fix?: Fix;
};
```

```ts
export type Problem = {
	/* … */
	fix?: () => void | never;
	editInfo?: EditInfo;
};
```

```ts
export type Problem = {
	/* … */
	fix?: (apply: boolean) => EditInfo;
}
```

```ts
export type Problem = {
	/* … */
	fix?: (doNotApply?: boolean) => EditInfo;
}
```

```ts
export type Problem = {
	/* … */
	fix?: () => EditInfo & { mutate: () => void | never };
}
```

```ts
export type Problem = {
	/* … */
	fix?: (Fixer) => FixingInstance;
}
```

`fix` must be run for the last 4 proposals, not for the first 2.

**Comment by Mouvedia** [comment]
> I think the fixed flag is mainly ambiguous in its meaning

I am not really sure what you are finding ambiguous so I am gonna try to make a summary of the situation.

The core rules which have the sentence

> can automatically fix **some** of the problems

in their README don't run `fix` for their unfixable problems.

e.g.

https://github.com/stylelint/stylelint/blob/3751cfb4c6ba8c94dbd88db75ff6bd77a1aa70bc/lib/rules/color-function-notation/index.mjs#L110

https://github.com/stylelint/stylelint/blob/3751cfb4c6ba8c94dbd88db75ff6bd77a1aa70bc/lib/rules/media-feature-range-notation/index.mjs#L101

We also discussed the scenari where `fix` would throw and it was decided that it couldn't be handled.
You were the one who, correctly, said that we should assume that `fix` cannot `throw`.
i.e. `fix` throwing is not handled

If you were not referencing these rare cases that may have impacted `fixed` accuracy it means that you are implying that plugin authors are erroneously putting their logic regarding mutation directly in their `fix` callback.
i.e. we shouldn't assume that the `fix` callback has mutated anything
Id expect 2 main ways to remediate this problem: events or diffing.

If I am correct in this assumption, I can see the connection if you are planning to call `fix` for #7192 regardless of the _appliance state_.
e.g. one of the last 4 proposals of https://github.com/stylelint/stylelint/issues/8233#issuecomment-2563071211

> This also requires _refactoring_ the verbose formatter so that it simply reports all fixers data.
> […]
> I do think there might a need for several other flags

If we do have the necessary flags implemented beforehand, why not.
Would it use `applied`?

**Comment by romainmenke** [comment]
I am unsure if we are talking about the same things :)

But maybe you are seeing a difference between these scenarios:
1. an error was encountered by a rule that can not fix anything, it doesn't have a `fix()` implementation.
2. an error was encountered by a rule that could fix some problems, but not all, it depends on the specific issue found in the source code
3. an error was encountered by a rule that could fix a specific problem but fixing was prevented by configuration comments, styleling config, ...
4. an error was encountered by a rule and it was fixed

My thinking is that only `4` is relevant and that we shouldn't record the other cases.

------

> We also discussed the scenari where fix would throw and it was decided that it couldn't be handled.
> You were the one who, correctly, said that we should assume that fix cannot throw.
> i.e. fix throwing is not handled

I don't see how that is relevant? 😕 


----

> If you were not referencing these rare cases that may have impacted fixed accuracy it means that you are implying that plugin authors are erroneously putting their logic regarding mutation directly in their fix callback.
> i.e. we shouldn't assume that the fix callback has mutated anything
> Id expect 2 main ways to remediate this problem: events or diffing.

I don't follow 😕 

The current implementation records `fixersData` with a flag `fixed` and adds entries even when no fix was applied.

I am suggesting to remove that flag and to no longer add entries unless an actual fix was applied. In other words, we only add an entry to `fixersData` in those cases where the flag would be `true` in the current implementation. The presence of the entry is sufficient indication that a fix happened. Entries where nothing was fixed are not useful imho.

I don't think this impacts the accuracy or correctness of the `fixersData` in as it is used today. The only use they have is for the verbose formatter and there we filter out any entries where `fixed` is `false`. So making this change wouldn't change anything.

----

Would it be more helpful if I open a patch?
Are there future plans you had for the `fixersData` which aren't knowable to me?

**Comment by Mouvedia** [comment]
Thanks, your explanation cleared it up.

**Comment by Mouvedia** [comment]
> Are there future plans you had for the `fixersData` which aren't knowable to me?

Not really unknowable; it's hinted by

https://github.com/stylelint/stylelint/blob/df2c64f4c83ab7b1269fe7367a5ecdd8fb7d0b39/lib/utils/report.mjs#L243

i.e. we want to eventually add the fixed ranges to the output of the `verbose` formatter to fulfil #4664 request

**Comment by romainmenke** [comment]
`range` is kept and is not affected by the removal of `fixed`. 

Also see: https://github.com/stylelint/stylelint/issues/4664#issuecomment-2568437149
